pkg/webhook/external/generic_list_options.go

```diff
@@ -0,0 +1,184 @@
+package external
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/pkg/expressions"
+)
+
+func listTargetObjects(
+	ctx context.Context,
+	c client.Client,
+	project string,
+	target kargoapi.GenericWebhookTarget,
+	actionEnv map[string]any,
+) ([]client.Object, error) {
+	var list client.ObjectList
+	switch target.Kind {
+	case kargoapi.GenericWebhookTargetKindWarehouse:
+		list = &kargoapi.WarehouseList{}
+	default:
+		return nil, fmt.Errorf("unsupported target kind: %q", target.Kind)
+	}
+	listOpts, err := buildListOptionsForTarget(project, target, actionEnv)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build list options: %w", err)
+	}
+	if err := c.List(ctx, list, listOpts...); err != nil {
+		return nil, fmt.Errorf("error listing %s targets: %w", target.Kind, err)
+	}
+	return listToObjects(list), nil
```

Comment:
```suggestion
	listOpts, err := buildListOptionsForTarget(project, target, actionEnv)
	if err != nil {
		return nil, fmt.Errorf("failed to build list options: %w", err)
	}

	switch target.Kind {
	case kargoapi.GenericWebhookTargetKindWarehouse:
		warehouses := &kargoapi.WarehouseList{}
		if err := c.List(ctx, warehouses, listOpts...); err != nil {
			return nil, fmt.Errorf("error listing %s targets: %w", target.Kind, err)
		}
		return itemsToObjects[kargoapi.Warehouse](warehouses.Items), nil
	default:
		return nil, fmt.Errorf("unsupported target kind: %q", target.Kind)
	}
```

Goes paired with the comment below.